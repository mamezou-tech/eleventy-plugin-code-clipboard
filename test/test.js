const plugin = require('..');

const mockAddShortcode = jest.fn();
const eleventyConfig = { addAsyncShortcode: mockAddShortcode };

describe('clipboard.js initialization', () => {
  afterEach(() => {
    mockAddShortcode.mockClear();
  });

  test('default plugin config', async () => {
    plugin.configFunction(eleventyConfig);
    expect(mockAddShortcode).toHaveBeenCalledTimes(1);
    const onloadScript = await mockAddShortcode.mock.calls[0][1]();
    expect(onloadScript).toMatchSnapshot();
  });

  test('custom plugin config', async () => {
    plugin.configFunction(eleventyConfig, {
      clipboardJSVersion: '2.0.0',
      buttonClass: 'test-button',
      successMessage: 'Success!!',
      failureMessage: 'Failure!!',
    });
    expect(mockAddShortcode).toHaveBeenCalledTimes(1);
    const onloadScript = await mockAddShortcode.mock.calls[0][1]();
    expect(onloadScript).toMatchSnapshot();
  });
});

describe('custom renderer', () => {
  let md;
  const mockRenderAttrs = jest.fn().mockReturnValue('class="test"');

  beforeEach(() => {
    md = {
      renderer: {
        rules: {
          fence: () => '<pre class="language-yaml"><code>test: unit-test<br />type: 11ty</code></pre>',
        },
      },
    };
  });

  afterEach(() => {
    mockRenderAttrs.mockClear();
  });

  test('default renderer config', () => {
    plugin.configFunction(eleventyConfig);
    plugin.markdownItCopyButton(md);
    const token = { tag: 'code', info: 'yaml', content: 'test: unit-test\ntype: 11ty' };
    const html = md.renderer.rules.fence(
      [token],
      0,
      undefined,
      undefined,
      { renderAttrs: mockRenderAttrs },
    );
    expect(html).toMatchSnapshot();
  });

  test('custom renderer config', () => {
    plugin.configFunction(eleventyConfig, {
      buttonClass: 'test-button',
    });
    const customRendererConfig = {
      iconifyUrl: 'https://example.com/custom.svg',
      iconStyle: 'width: 24px; height: 24px;',
      iconClass: 'test-copy',
      iconTag: 'test-tag',
      buttonClass: 'test-button',
      buttonStyle: 'opacity: 0.1;',
      additionalButtonClass: 'another-button-class',
      title: 'test-title',
    };
    plugin.markdownItCopyButton(md, customRendererConfig);
    const token = { tag: 'code', info: 'yaml', content: 'test: unit-test\ntype: 11ty' };
    const html = md.renderer.rules.fence(
      [token],
      0,
      undefined,
      undefined,
      { renderAttrs: mockRenderAttrs },
    );
    expect(html).toMatchSnapshot();
  });

  test('non-lang-code token(not applied clipboard)', () => {
    plugin.configFunction(eleventyConfig);
    plugin.markdownItCopyButton(md);
    const token = { tag: 'code', info: undefined, content: 'article title' };
    const html = md.renderer.rules.fence([token], 0);
    expect(html).toMatchSnapshot();
  });
});
