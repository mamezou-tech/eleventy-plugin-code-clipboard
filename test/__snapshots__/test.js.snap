// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`clipboard.js initialization custom plugin config 1`] = `
"<script>function showTooltip(o,e){o.trigger.className.includes(\\"tooltipped\\")||(o.trigger.children[0].className=\\"tooltipped tooltipped-s\\",o.trigger.children[0].ariaLabel=e)}window.addEventListener(\\"load\\",()=>{const o=new ClipboardJS(\\".test-button\\");o.on(\\"success\\",o=>showTooltip(o,\\"Success!!\\")),o.on(\\"error\\",o=>showTooltip(o,\\"Failure!!\\"))});</script>
<script async src=\\"https://cdn.jsdelivr.net/npm/clipboard@2.0.0/dist/clipboard.min.js\\"></script>"
`;

exports[`clipboard.js initialization default plugin config 1`] = `
"<script>function showTooltip(o,e){o.trigger.className.includes(\\"tooltipped\\")||(o.trigger.children[0].className=\\"tooltipped tooltipped-s\\",o.trigger.children[0].ariaLabel=e)}window.addEventListener(\\"load\\",()=>{const o=new ClipboardJS(\\".code-copy\\");o.on(\\"success\\",o=>showTooltip(o,\\"Copied!\\")),o.on(\\"error\\",o=>showTooltip(o,\\"Failed...\\"))});</script>
<script async src=\\"https://cdn.jsdelivr.net/npm/clipboard@2.0.11/dist/clipboard.min.js\\"></script>"
`;

exports[`custom renderer custom renderer config 1`] = `
"
<div style=\\"position: relative\\">
  <pre class=\\"language-yaml\\"><code id=\\"code-0\\">test: unit-test<br />type: 11ty</code></pre>
  <button class=\\"test-button another-button-class\\"
    data-clipboard-target=\\"#code-0\\"
    style=\\"opacity: 0.1;\\" title=\\"test-title\\">
    <test-tag style=\\"display:inline-block;background:url(https://example.com/custom.svg) no-repeat center center / contain;width: 24px; height: 24px;\\" class=\\"test-copy\\"></test-tag>
  </button>
</div>
"
`;

exports[`custom renderer default renderer config 1`] = `
"
<div style=\\"position: relative\\">
  <pre class=\\"language-yaml\\"><code id=\\"code-0\\">test: unit-test<br />type: 11ty</code></pre>
  <button class=\\"code-copy \\"
    data-clipboard-target=\\"#code-0\\"
    style=\\"position: absolute; top: 7.5px; right: 6px; padding-top: 3px; cursor: pointer; outline: none; opacity: 0.8;\\" title=\\"Copy\\">
    <span style=\\"display:inline-block;background:url(https://api.iconify.design/mdi/content-copy.svg) no-repeat center center / contain;width: 16px; height: 16px;\\" class=\\"\\"></span>
  </button>
</div>
"
`;

exports[`custom renderer non-lang-code token(not applied clipboard) 1`] = `"<pre class=\\"language-yaml\\"><code>test: unit-test<br />type: 11ty</code></pre>"`;
